### PR TITLE
operator to use fabric-config

### PR DESCRIPTION
**Pull request by suryalnvs**
Operator to use fabric-config for configuration generation

Signed-off-by: Surya Lanka <suryalnvs@gmail.com>

### CODE REVIEW

File: tools/operator/fabricconfiguration/configtx.go [hunk 1]
```diff
@@ -0,0 +1,933 @@
...
+		}
+		anchorPeers = append(anchorPeers, anchorPeer)
+		peerOrganization := &genesisconfig.Organization{
+			Name: org.Name,
+			ID: org.MSPID,
+			MSPDir: paths.JoinPath(peerOrgsPath, fmt.Sprintf("%s/msp/", org.Name)),
+			Policies: map[string]*genesisconfig.Policy{
+				configtx.ReadersPolicyKey: {
+					Type: configtx.SignaturePolicyType,
+					Rule: "OR('" + org.MSPID + ".admin', '" + org.MSPID + ".peer', '" + org.MSPID + ".client')",
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Replace with `fmt.Sprintf`

File: tools/operator/fabricconfiguration/configtx.go [hunk 2]
```diff
@@ -0,0 +1,933 @@
...
+				},
+				configtx.WritersPolicyKey: {
+					Type: configtx.SignaturePolicyType,
+					Rule: "OR('" + org.MSPID + ".admin', '" + org.MSPID + ".client')",
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Replace with `fmt.Sprintf`

File: tools/operator/fabricconfiguration/configtx.go [hunk 3]
```diff
@@ -0,0 +1,933 @@
...
+				},
+				configtx.AdminsPolicyKey: {
+					Type: configtx.SignaturePolicyType,
+					Rule: "OR('" + org.MSPID + ".admin')",
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Replace with `fmt.Sprintf`

File: tools/operator/fabricconfiguration/configtx.go [hunk 4]
```diff
@@ -0,0 +1,933 @@
...
+				},
+				configtx.EndorsementPolicyKey: {
+					Type: configtx.SignaturePolicyType,
+					Rule: "OR('" + org.MSPID + ".peer')",
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Replace with `fmt.Sprintf`

File: tools/operator/fabricconfiguration/configtx.go [hunk 5]
```diff
@@ -0,0 +1,933 @@
...
+		},
+	}
+	return configtxConfiguration.Profiles[profile]
+}
+
+// The ProviderType of a member relative to the member API
+const (
+	FABRIC ProviderType = iota // MSP is of FABRIC type
+	IDEMIX                     // MSP is of IDEMIX type
+	OTHER                      // MSP is of OTHER TYPE
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Unused, delete

File: tools/operator/fabricconfiguration/configtx.go [hunk 6]
```diff
@@ -0,0 +1,933 @@
...
+
+	// NOTE: as new types are added to this set,
+	// the mspTypes map below must be extended
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Remove stray comments

File: tools/operator/fabricconfiguration/configtx.go [hunk 7]
```diff
@@ -0,0 +1,933 @@
...
+	size = strings.Trim(size, " MB") 
+	sizeToInt, _ = strconv.Atoi(size)
+	return uint32(sizeToInt * 1024 * 1024)
+}
+
+func doOutputBlock(config *genesisconfig.Profile, channelID string, outputBlock string) error {
+	channel := newChannel(config)
+	genesisBlock, err := configtx.NewSystemChannelGenesisBlock(channel, channelID)
+	if err != nil {
+		panic(err)
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Return error, don't panic

File: tools/operator/fabricconfiguration/configtx.go [hunk 8]
```diff
@@ -0,0 +1,933 @@
...
+}
+
+func doOutputChannelCreateTx(conf *genesisconfig.Profile, channelID string, outputChannelCreateTx string) error {
+	logger.Info("Generating new channel configtx")
+
+	channel := newChannel(conf)
+	configUpdate, err := configtx.NewMarshaledCreateChannelTx(channel, channelID)
+	env, err := configtx.NewEnvelope(configUpdate)
+	if err != nil {
+		panic(err)
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Return error

File: tools/operator/fabricconfiguration/configtx.go [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func newChannel(baseProfile *genesisconfig.Profile) configtx.Channel {
+	// Application section
+	application := configtx.Application{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 10]
```diff
@@ -0,0 +1,933 @@
...
+	if baseProfile.Application != nil {
+		appOrgs := newOrganization(baseProfile.Application.Organizations)
+		appCapabilities := []string{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 11]
```diff
@@ -0,0 +1,933 @@
...
+		application = configtx.Application{
+			Organizations: appOrgs,
+			Capabilities:  appCapabilities,
+			Policies:      appPolicies,
+			ACLs:          baseProfile.Application.ACLs,
+		}
+	}
+
+	// Orderer section
+	o := configtx.Orderer{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 12]
```diff
@@ -0,0 +1,933 @@
...
+	if baseProfile.Orderer != nil {
+		addresses := []configtx.Address{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 13]
```diff
@@ -0,0 +1,933 @@
...
+		for _, addr := range baseProfile.Orderer.Addresses {
+			addrslice := strings.Split(addr, ":")
+			port, err := strconv.Atoi(addrslice[1])
+			if err != nil {
+				panic(err)
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Return error.

All places that panic should return an error instead

File: tools/operator/fabricconfiguration/configtx.go [hunk 14]
```diff
@@ -0,0 +1,933 @@
...
+			}
+			addresses = append(addresses, configtx.Address{
+				Host: addrslice[0],
+				Port: port,
+			})
+		}
+		kafka := orderer.Kafka{
+			Brokers: baseProfile.Orderer.Kafka.Brokers,
+		}
+		consentors := []orderer.Consenter{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 15]
```diff
@@ -0,0 +1,933 @@
...
+		if err != nil {
+			return certificateList, err
+		}
+		certificateList = append(certificateList, certificate)
+	}
+	return certificateList, nil
+}
+
+func parseCRL(crls [][]byte) ([]*pkix.CertificateList, error) {
+	certificateLists := []*pkix.CertificateList{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 16]
```diff
@@ -0,0 +1,933 @@
...
+	pemBlock, _ := pem.Decode(priv)
+	privateKey, err := x509.ParsePKCS8PrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing PKCS#8 private key: %v", err)
+	}
+	return privateKey, nil
+}
+
+func parseOUIdentifiers(identifiers []*msp.FabricOUIdentifier) ([]membership.OUIdentifier, error) {
+	fabricIdentifiers := []membership.OUIdentifier{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 17]
```diff
@@ -0,0 +1,933 @@
...
+			Certificate:                  cert,
+			OrganizationalUnitIdentifier: identifier.OrganizationalUnitIdentifier,
+		}
+		fabricIdentifiers = append(fabricIdentifiers, fabricOUIdentifier)
+	}
+	return fabricIdentifiers, nil
+}
+
+func getMSPConfig(config *msp.MSPConfig) (configtx.MSP, error) {
+	fabricMSPConfig := &msp.FabricMSPConfig{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go:751]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 18]
```diff
@@ -0,0 +1,933 @@
...
+	admincertDir := filepath.Join(dir, admincerts)
+	intermediatecertsDir := filepath.Join(dir, intermediatecerts)
+	crlsDir := filepath.Join(dir, crlsfolder)
+	configFile := filepath.Join(dir, configfilename)
+	tlscacertDir := filepath.Join(dir, tlscacerts)
+	tlsintermediatecertsDir := filepath.Join(dir, tlsintermediatecerts)
+
+	cacerts, err := getPemMaterialFromDir(cacertDir)
+	if err != nil || len(cacerts) == 0 {
+		return nil, errors.WithMessagef(err, "could not load a valid ca certificate from directory %s", cacertDir)
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Don't use the error's package, use `fmt.Errorf` to construct errors. There are several places we use it today

File: tools/operator/fabricconfiguration/configtx.go [hunk 19]
```diff
@@ -0,0 +1,933 @@
...
+		if err != nil {
+			continue
+		}
+		content = append(content, item)
+	}
+	return content, nil
+}
+
+func newPolicies(policies map[string]*genesisconfig.Policy) map[string]configtx.Policy {
+	txpolicies := map[string]configtx.Policy{}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go:1066]: Use `var` declaration

File: tools/operator/fabricconfiguration/configtx.go [hunk 20]
```diff
@@ -0,0 +1,933 @@
...
+		}
+	}
+
+	if config.OutputChannelCreateTx != "" {
+		if err := doOutputChannelCreateTx(profileConfig, config.ChannelID, config.OutputChannelCreateTx); err != nil {
+			logger.Fatalf("Error on outputChannelCreateTx: %s", err)
+		}
+	}
+	return nil
+}
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/configtx.go]: Newline missing

File: tools/operator/fabricconfiguration/genesisconfig/config.go [hunk 21]
```diff
@@ -0,0 +1,448 @@
+/*
```
Comment (lindluni) [inline tools/operator/fabricconfiguration/genesisconfig/config.go]: All of the functions in this file can be removed, we only use the structs, gut it of all functions.

**Review by lindluni — CHANGES_REQUESTED** [review]

File: tools/operator/fabricconfiguration/configtx.go (hunk 1 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 2 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 3 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 4 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 5 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 6 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 7 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 8 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 9 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 10 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 11 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 12 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 13 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 14 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 15 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 16 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 18 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 19 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go:1066]: Fixed

File: tools/operator/fabricconfiguration/configtx.go (hunk 20 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/configtx.go]: Fixed

File: tools/operator/fabricconfiguration/genesisconfig/config.go (hunk 21 above)
Comment (suryalnvs) [reply tools/operator/fabricconfiguration/genesisconfig/config.go]: removed the file.

**Review by lindluni — APPROVED** [review]

**Pull request merged**